pytorch_lightning/overrides/distributed.py

```diff
@@ -75,3 +76,58 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
             model.reducer.prepare_for_backward([])
     else:
         model.require_forward_param_sync = False
+
+
+# Taken from https://github.com/jpuigcerver/PyLaia/blob/master/laia/data/unpadded_distributed_sampler.py#L35
+class UnRepeatedDistributedSampler(DistributedSampler):
```

Comment:
does prediction really need this? is it reasonable for the user to identify the batch based on the batch/batch idx/dataloader idx? could this lead to more confusion for differences between batch_idx and batch_indices? does this only work for certain types of datasets?

Reply:
Yes, this is required, as we don't want to lose any sample nor repeat last_batch on all ranks and perform extra computation.

Reply:
The batch_indexes are the element of truth for the tracking.
I don't expect people to use batch_idx and it should be fine to have un-balanced number of batches on all ranks.
We won't use this for IterableDataset, as we don't know their length. Same for IndexBatchSampler.